# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [114]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, StandardScaler

In [116]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

**Check the shape of your data**

In [118]:
spaceship.shape

(8693, 14)

**Check for data types**

In [120]:
spaceship.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

**Check for missing values**

In [122]:
missing_values = spaceship.isnull().sum()
print(missing_values)

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64


There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [124]:
spaceship_cleaned = spaceship.dropna()
spaceship_cleaned

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B/0/P    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F/0/S    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A/0/S    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A/0/S    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F/1/S    TRAPPIST-1e  16.0  False   
...          ...        ...       ...       ...            ...   ...    ...   
8688     9276_01     Europa     False    A/98/P    55 Cancri e  41.0   True   
8689     9278_01      Earth      True  G/1499/S  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False  G/1500/S    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S    55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0             0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1           109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2            43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3             0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4           303.0       70.0         151.0   565.0     2.0  Willy Santantines   
...           ...        ...           ...     ...     ...                ...   
8688          0.0     6819.0           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0        0.0           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0        0.0        1872.0     1.0     0.0       Fayey Connon   
8691          0.0     1049.0           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0     4688.0           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
0           False  
1            True  
2           False  
3           False  
4            True  
...           ...  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True  

[6606 rows x 14 columns]

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [126]:
spaceship_cleaned['Cabin'] = spaceship_cleaned['Cabin'].str[0]
print(spaceship_cleaned[['Cabin']].head())

  Cabin
0     B
1     F
2     A
3     A
4     F


/var/folders/43/30gtvzf54299gghzdyc8_l900000gn/T/ipykernel_25978/2081853163.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spaceship_cleaned['Cabin'] = spaceship_cleaned['Cabin'].str[0]


- Drop PassengerId and Name

In [128]:
spaceship_cleaned = spaceship_cleaned.drop(columns=['PassengerId', 'Name'])
spaceship_cleaned

HomePlanet CryoSleep Cabin    Destination   Age    VIP  RoomService  \
0        Europa     False     B    TRAPPIST-1e  39.0  False          0.0   
1         Earth     False     F    TRAPPIST-1e  24.0  False        109.0   
2        Europa     False     A    TRAPPIST-1e  58.0   True         43.0   
3        Europa     False     A    TRAPPIST-1e  33.0  False          0.0   
4         Earth     False     F    TRAPPIST-1e  16.0  False        303.0   
...         ...       ...   ...            ...   ...    ...          ...   
8688     Europa     False     A    55 Cancri e  41.0   True          0.0   
8689      Earth      True     G  PSO J318.5-22  18.0  False          0.0   
8690      Earth     False     G    TRAPPIST-1e  26.0  False          0.0   
8691     Europa     False     E    55 Cancri e  32.0  False          0.0   
8692     Europa     False     E    TRAPPIST-1e  44.0  False        126.0   

      FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0           0.0           0.0     0.0     0.0        False  
1           9.0          25.0   549.0    44.0         True  
2        3576.0           0.0  6715.0    49.0        False  
3        1283.0         371.0  3329.0   193.0        False  
4          70.0         151.0   565.0     2.0         True  
...         ...           ...     ...     ...          ...  
8688     6819.0           0.0  1643.0    74.0        False  
8689        0.0           0.0     0.0     0.0        False  
8690        0.0        1872.0     1.0     0.0         True  
8691     1049.0           0.0   353.0  3235.0        False  
8692     4688.0           0.0     0.0    12.0         True  

[6606 rows x 12 columns]

- For non-numerical columns, do dummies.

In [130]:
object_df = spaceship_cleaned.select_dtypes("object")
object_df

HomePlanet CryoSleep Cabin    Destination    VIP
0        Europa     False     B    TRAPPIST-1e  False
1         Earth     False     F    TRAPPIST-1e  False
2        Europa     False     A    TRAPPIST-1e   True
3        Europa     False     A    TRAPPIST-1e  False
4         Earth     False     F    TRAPPIST-1e  False
...         ...       ...   ...            ...    ...
8688     Europa     False     A    55 Cancri e   True
8689      Earth      True     G  PSO J318.5-22  False
8690      Earth     False     G    TRAPPIST-1e  False
8691     Europa     False     E    55 Cancri e  False
8692     Europa     False     E    TRAPPIST-1e  False

[6606 rows x 5 columns]

In [134]:
num_df= spaceship_cleaned.select_dtypes("number")
num_df

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck
0     39.0          0.0        0.0           0.0     0.0     0.0
1     24.0        109.0        9.0          25.0   549.0    44.0
2     58.0         43.0     3576.0           0.0  6715.0    49.0
3     33.0          0.0     1283.0         371.0  3329.0   193.0
4     16.0        303.0       70.0         151.0   565.0     2.0
...    ...          ...        ...           ...     ...     ...
8688  41.0          0.0     6819.0           0.0  1643.0    74.0
8689  18.0          0.0        0.0           0.0     0.0     0.0
8690  26.0          0.0        0.0        1872.0     1.0     0.0
8691  32.0          0.0     1049.0           0.0   353.0  3235.0
8692  44.0        126.0     4688.0           0.0     0.0    12.0

[6606 rows x 6 columns]

In [136]:
from sklearn.preprocessing import OneHotEncoder
pd.get_dummies(object_df, dtype= int)

HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  \
0                    0                  1                0                1   
1                    1                  0                0                1   
2                    0                  1                0                1   
3                    0                  1                0                1   
4                    1                  0                0                1   
...                ...                ...              ...              ...   
8688                 0                  1                0                1   
8689                 1                  0                0                0   
8690                 1                  0                0                1   
8691                 0                  1                0                1   
8692                 0                  1                0                1   

      CryoSleep_True  Cabin_A  Cabin_B  Cabin_C  Cabin_D  Cabin_E  Cabin_F  \
0                  0        0        1        0        0        0        0   
1                  0        0        0        0        0        0        1   
2                  0        1        0        0        0        0        0   
3                  0        1        0        0        0        0        0   
4                  0        0        0        0        0        0        1   
...              ...      ...      ...      ...      ...      ...      ...   
8688               0        1        0        0        0        0        0   
8689               1        0        0        0        0        0        0   
8690               0        0        0        0        0        0        0   
8691               0        0        0        0        0        1        0   
8692               0        0        0        0        0        1        0   

      Cabin_G  Cabin_T  Destination_55 Cancri e  Destination_PSO J318.5-22  \
0           0        0                        0                          0   
1           0        0                        0                          0   
2           0        0                        0                          0   
3           0        0                        0                          0   
4           0        0                        0                          0   
...       ...      ...                      ...                        ...   
8688        0        0                        1                          0   
8689        1        0                        0                          1   
8690        1        0                        0                          0   
8691        0        0                        1                          0   
8692        0        0                        0                          0   

      Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                           1          1         0  
1                           1          1         0  
2                           1          0         1  
3                           1          1         0  
4                           1          1         0  
...                       ...        ...       ...  
8688                        0          0         1  
8689                        0          1         0  
8690                        1          1         0  
8691                        0          1         0  
8692                        1          1         0  

[6606 rows x 18 columns]

**Perform Train Test Split**

In [138]:
enc = OneHotEncoder()

In [140]:
enc.fit(object_df)

OneHotEncoder()

In [142]:
enc.transform(object_df).toarray()

array([[0., 1., 0., ..., 1., 1., 0.],
       [1., 0., 0., ..., 1., 1., 0.],
       [0., 1., 0., ..., 1., 0., 1.],
       ...,
       [1., 0., 0., ..., 1., 1., 0.],
       [0., 1., 0., ..., 0., 1., 0.],
       [0., 1., 0., ..., 1., 1., 0.]])

In [144]:
features = spaceship_cleaned.drop(columns=['HomePlanet', 'CryoSleep', 'Cabin', 'VIP', 'Destination', 'Transported'])
features

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck
0     39.0          0.0        0.0           0.0     0.0     0.0
1     24.0        109.0        9.0          25.0   549.0    44.0
2     58.0         43.0     3576.0           0.0  6715.0    49.0
3     33.0          0.0     1283.0         371.0  3329.0   193.0
4     16.0        303.0       70.0         151.0   565.0     2.0
...    ...          ...        ...           ...     ...     ...
8688  41.0          0.0     6819.0           0.0  1643.0    74.0
8689  18.0          0.0        0.0           0.0     0.0     0.0
8690  26.0          0.0        0.0        1872.0     1.0     0.0
8691  32.0          0.0     1049.0           0.0   353.0  3235.0
8692  44.0        126.0     4688.0           0.0     0.0    12.0

[6606 rows x 6 columns]

In [146]:
target = spaceship_cleaned['Transported']
target

0       False
1        True
2       False
3       False
4        True
        ...  
8688    False
8689    False
8690     True
8691    False
8692     True
Name: Transported, Length: 6606, dtype: bool

In [148]:
features_num = pd.get_dummies(features)

In [162]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features_num, target, random_state = 42)

In [164]:
X_train

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck
4737  42.0          0.0        0.0           0.0     0.0     0.0
1688  33.0          0.0        0.0           0.0     0.0     0.0
3205  29.0          0.0        4.0           0.0     0.0   904.0
5859  27.0         55.0        0.0          50.0  1849.0     1.0
3956  32.0       1331.0      416.0           0.0   665.0    64.0
...    ...          ...        ...           ...     ...     ...
4984  31.0          0.0        0.0           0.0     0.0     0.0
6864  22.0         28.0      194.0           2.0     0.0   370.0
6919  25.0          0.0      115.0           0.0  1823.0  4041.0
7137  40.0          0.0       21.0           0.0  1503.0  1392.0
1162  27.0          0.0        0.0         863.0    33.0     2.0

[4954 rows x 6 columns]

**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [170]:
scaler = MinMaxScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [172]:
from sklearn.neighbors import KNeighborsRegressor

In [176]:
knn = KNeighborsRegressor(n_neighbors=3)
knn.fit(X_train_scaled, y_train)

KNeighborsRegressor(n_neighbors=3)

- Evaluate your model's performance. Comment it

In [178]:
knn.score(X_test_scaled, y_test)

0.19377393217714467

Score of 0.1938 suggests that your model has relatively low predictive accuracy.